src/main.cpp

```diff
@@ -6355,7 +6355,25 @@ bool ProcessMessage(CNode *pfrom, std::string strCommand, CDataStream &vRecv, in
 
         bool fCanDirectFetch = CanDirectFetch(chainparams.GetConsensus());
         CNodeState *nodestate = State(pfrom->GetId());
-        nodestate->fFirstHeadersReceived = true;
+
+        // During the initial peer handshake we must receive the initial headers which should be greater
+        // than or equal to our block height at the time of requesting GETHEADERS. This is because the peer has
+        // advertised a height >= to our own. Furthermore, because the headers max returned is as much as 2000 this
+        // could not be a mainnet re-org.
+        if (!nodestate->fFirstHeadersReceived)
+        {
+            // We want to make sure that the peer doesn't just send us any old valid header. The last header block
+            // hash they send us should now be equal to our current pindexLast block hash.
+            if (hashLastBlock == pindexLast->GetBlockHash())
+            {
```

Comment:
Discussed in private with @ptschip.  I don't believe pIndexLast can ever be null (AcceptBlockHeader doesn't permit it) which makes some of the other checks above this code redundant.  But I do believe that equality will always be true; pIndexLast points to the header passed to AcceptBlockHeader, not the chain tip.